File: macos/Tests/TCShellCoreTests/HarnessSurfaceTests.swift
import XCTest

@testable import TCShellCore

/// The harness list's decoding and its branches, without the dylib.
///
/// The injected calls deliberately do NOT reimplement the Rust branch
/// tables: a fake that reproduced the real mapping would let this suite pass
/// while the shell had stopped asking the shared table at all. Every one of
/// them is a spy or a stub returning a value the test chose.
final class HarnessSurfaceTests: XCTestCase {
    private let payload = """
        {"destination":"DESTINATION","subtitle":"SUBTITLE",
         "offer_title":"T","offer_what":"WHAT","offer_exposure":"EXPOSURE",
         "offer_no_repoint":"NO-REPOINT","offer_accept":"ACCEPT",
         "offer_decline":"DECLINE","offer_asked_once":"ONCE",
         "settings_title":"S-TITLE","settings_toggle":"S-TOGGLE",
         "settings_applies_at_once":"S-AT-ONCE","state_off":"S-OFF","state_unknown":"S-UNKNOWN","state_unreported":"S-UNREPORTED","state_stopping":"S-STOPPING",
         "state_running":"S-RUNNING","state_running_no_backends":"S-NO-BACKENDS",
         "state_running_elsewhere":"S-ELSEWHERE","state_port_in_use":"S-PORT",
         "state_start_failed":"S-FAILED","state_crashed":"S-CRASHED",
         "quit_also_stops":"QUIT","write_unconfirmed":"UNCONFIRMED","settings_moved":"MOVED","tray_turn_off":"TRAYOFF","tray_open_to_turn_on":"TRAYON",
         "harnesses_title":"H-TITLE","harnesses_what":"H-WHAT",
         "harness_not_connected":"H-NOT-CONNECTED",
         "harness_connected_nothing_seen":"H-NOTHING-SEEN",
         "harness_answering":"H-ANSWERING","harness_connect":"H-CONNECT",
         "harness_disconnect":"H-DISCONNECT",
         "harness_preview_title":"H-PREVIEW","harness_preview_confirm":"H-CONFIRM",
         "harness_preview_cancel":"H-CANCEL","harness_slot_taken":"H-TAKEN",
         "harness_needs_restart":"H-RESTART","harnesses_none_found":"H-NONE",
         "harness_unreadable_config":"H-UNREADABLE"}
        """

    private func copy() -> PrivateInferenceCopy {
        guard let copy = PrivateInferenceCopy.decode(fromJSON: payload) else {
            XCTFail("the fixture payload must decode")
            fatalError("unreachable")
        }
        return copy
    }

    private static let oneRow = """
        {"catalog_present":false,"destination_port":8891,
         "harnesses":[{"id":"claude","name":"Claude Code","installed":true,
           "connected":false,"config_path":"/Users/x/.claude/settings.json",
           "connect_command":"tc connect claude","family":"anthropic",
           "state":"not_connected","last_call_at":null,
           "can_connect":true,"can_disconnect":false}],
         "activity":{"readable":true,"window_hours":24,"last_call_at":null,"families":[]}}
        """

    // MARK: - The list

    /// A row this build cannot read must not silently become a connected one.
    func testAMalformedPayloadYieldsNoRowsRatherThanAGuess() {
        let list = HarnessSurface.list(fromJSON: #"{"harnesses":[{"name":"x"}]}"#)
        XCTAssertEqual(list.harnesses.count, 0)
        XCTAssertNil(list.destinationPort)
    }

    /// The file path is part of the row, always. A tool nobody expected to be
    /// set up is a question about which file, every time.
    func testTheRowCarriesTheFileItWouldChange() {
        let list = HarnessSurface.list(fromJSON: Self.oneRow)
        XCTAssertEqual(list.harnesses.count, 1)
        XCTAssertEqual(list.harnesses[0].configPath, "/Users/x/.claude/settings.json")
        XCTAssertEqual(list.harnesses[0].connectCommand, "tc connect claude")
        XCTAssertEqual(list.destinationPort, 8891)
        XCTAssertFalse(list.catalogPresent)
    }

    /// A tool with nowhere to write is listed with a nil path rather than
    /// dropped: "we have never heard of it" and "we cannot find its file"
    /// are different answers.
    func testARowWithNoConfigPathStillDecodes() {
        let json = Self.oneRow.replacingOccurrences(
            of: "\"config_path\":\"/Users/x/.claude/settings.json\"", with: "\"config_path\":null")
        let list = HarnessSurface.list(fromJSON: json)
        XCTAssertEqual(list.harnesses.count, 1)
        XCTAssertNil(list.harnesses[0].configPath)
    }

    // MARK: - The state, and what may be painted as working

    /// The state code is asked of the shared table, never matched on here.
    func testTheStateCodeIsAskedOfTheSharedTable() {
        let seen = SpyBox()
        let calls = HarnessCalls(
            stateCode: { seen.record($0); return 33 },
            planOutcomeCode: { _ in 40 },
            actionAvailable: { _, _, _ in false })
        XCTAssertEqual(HarnessSurface.state("answering", calls: calls), .answering)
        XCTAssertEqual(seen.values, ["answering"])
    }

    /// A state this build has no words for is not painted as working, and a
    /// tone table that answered `.clear` for it would be the fail-open the
    /// whole surface exists to prevent.
    func testOnlyAnsweringReadsAsWorking() {
        XCTAssertTrue(HarnessSurface.tone(.answering).readsAsWorking)
        for state in [
            HarnessState.notConnected, .connectedNoCalls, .activityShared, .unknown,
        ] {
            XCTAssertFalse(
                HarnessSurface.tone(state).readsAsWorking,
                "\(state) must not be painted as working")
        }
    }

    /// "One of these two answered" is not "this one is answering", so the
    /// shared-activity state borrows neither the answering sentence nor the
    /// nothing-seen one -- it says nothing, which is all it can honestly say.
    func testTheUnattributableStatesClaimNothing() {
        XCTAssertEqual(HarnessSurface.stateSentence(.notConnected, copy: copy()), "H-NOT-CONNECTED")
        XCTAssertEqual(
            HarnessSurface.stateSentence(.connectedNoCalls, copy: copy()), "H-NOTHING-SEEN")
        XCTAssertEqual(HarnessSurface.stateSentence(.answering, copy: copy()), "H-ANSWERING")
        XCTAssertNil(HarnessSurface.stateSentence(.activityShared, copy: copy()))
        XCTAssertNil(HarnessSurface.stateSentence(.unknown, copy: copy()))
    }

    /// The running copy of a tool read its settings when it started. The
    /// sentence about that stays up while the file says one thing and no
    /// call has been attributed, and goes the moment one is.
    func testTheRestartSentenceStaysUpUntilACallIsAttributed() {
        let connected = Self.oneRow.replacingOccurrences(
            of: "\"connected\":false", with: "\"connected\":true")
        let row = HarnessSurface.list(fromJSON: connected).harnesses[0]
        XCTAssertEqual(
            HarnessSurface.restartSentence(row, state: .connectedNoCalls, copy: copy()), "H-RESTART")
        XCTAssertEqual(
            HarnessSurface.restartSentence(row, state: .activityShared, copy: copy()), "H-RESTART")
        XCTAssertNil(HarnessSurface.restartSentence(row, state: .answering, copy: copy()))

        let notConnected = HarnessSurface.list(fromJSON: Self.oneRow).harnesses[0]
        XCTAssertNil(
            HarnessSurface.restartSentence(notConnected, state: .notConnected, copy: copy()))
    }

    /// An unfamiliar code is `unknown`, never the value next to it.
    func testAnUnfamiliarStateCodeIsUnknown() {
        for code: Int32 in [0, 22, 29, 35, 99, -1] {
            XCTAssertEqual(HarnessState.fromABI(code), .unknown)
        }
    }

    // MARK: - The actions offered on a row

    /// Whether a button appears is the shared table's decision, and a shell
    /// that answered it from `installed` alone would hide the control that
    /// removes what we put in an uninstalled tool's file.
    func testTheActionsOfferedComeFromTheSharedTable() {
        let seen = SpyBox()
        let calls = HarnessCalls(
            stateCode: { _ in 31 },
            planOutcomeCode: { _ in 40 },
            actionAvailable: { action, _, _ in
                seen.record(action)
                return action == "disconnect"
            })
        let row = HarnessSurface.list(fromJSON: Self.oneRow).harnesses[0]
        XCTAssertFalse(HarnessSurface.canConnect(row, calls: calls))
        XCTAssertTrue(HarnessSurface.canDisconnect(row, calls: calls))
        XCTAssertEqual(seen.values, ["connect", "disconnect"])
    }

    // MARK: - The plan

    /// Only `changes` is committable, and a plan id is what makes the commit
    /// possible at all -- the shell cannot construct a write of its own.
    func testOnlyAChangesPlanWithAnIdIsCommittable() {
        let calls = HarnessCalls(
            stateCode: { _ in 31 },
            planOutcomeCode: { outcome in outcome == "changes" ? 41 : 42 },
            actionAvailable: { _, _, _ in true })
        let changes = HarnessSurface.plan(
            fromJSON: #"""
                {"id":"claude","action":"connect","outcome":"changes","plan_id":"c0ffee",
                 "path":"/p","changes":["set a thing"],"occupied":[]}
                """#)
        XCTAssertEqual(changes?.changes, ["set a thing"])
        XCTAssertEqual(changes.map { HarnessSurface.canCommit($0, calls: calls) }, true)

        let noop = HarnessSurface.plan(
            fromJSON: #"""
                {"id":"claude","action":"connect","outcome":"noop","plan_id":null,
                 "path":"/p","changes":[],"occupied":[]}
                """#)
        XCTAssertEqual(noop.map { HarnessSurface.outcome($0, calls: calls) }, .noop)
        XCTAssertEqual(noop.map { HarnessSurface.canCommit($0, calls: calls) }, false)
    }

    /// A committable outcome with no plan id is still not committable. The
    /// daemon mints the id; a shell that fell back to sending the tool id
    /// would have constructed a write.
    func testAChangesPlanWithoutAnIdIsNotCommittable() {
        let calls = HarnessCalls(
            stateCode: { _ in 31 }, planOutcomeCode: { _ in 41 },
            actionAvailable: { _, _, _ in true })
        let plan = HarnessSurface.plan(
            fromJSON: #"""
                {"id":"claude","action":"connect","outcome":"changes","plan_id":null,
                 "path":"/p","changes":["set a thing"],"occupied":[]}
                """#)
        XCTAssertEqual(plan.map { HarnessSurface.canCommit($0, calls: calls) }, false)
    }

    /// A file we refused to rewrite is not a file with nothing to change,
    /// and collapsing the two tells a contributor with a broken config that
    /// everything is fine.
    func testARefusedFileIsNotAFileWithNothingToChange() {
        let calls = HarnessCalls(
            stateCode: { _ in 31 },
            planOutcomeCode: { outcome in outcome == "unparseable" ? 43 : 42 },
            actionAvailable: { _, _, _ in true })
        let refused = HarnessSurface.plan(
            fromJSON: #"""
                {"id":"claude","action":"connect","outcome":"unparseable","plan_id":null,
                 "path":"/p","changes":[],"occupied":[]}
                """#)
        XCTAssertEqual(
            refused.map { HarnessSurface.outcomeSentence($0, copy: copy(), calls: calls) },
            "H-UNREADABLE")
        let noop = HarnessSurface.plan(
            fromJSON: #"""
                {"id":"claude","action":"connect","outcome":"noop","plan_id":null,
                 "path":"/p","changes":[],"occupied":[]}
                """#)
        XCTAssertEqual(
            noop.map { HarnessSurface.outcomeSentence($0, copy: copy(), calls: calls) }, .some(nil))
    }

    /// An occupied slot survives to the screen, never swallowed. This is the
    /// rule most likely to be lost to a well-meaning simplification.
    func testAnOccupiedSlotSurvivesToTheScreen() {
        let plan = HarnessSurface.plan(
            fromJSON: #"""
                {"id":"claude","action":"connect","outcome":"noop","plan_id":null,"path":"/p",
                 "changes":[],
                 "occupied":[{"slot":"env.ANTHROPIC_BASE_URL","current":"https://theirs.example"}]}
                """#)
        XCTAssertEqual(plan?.occupied.first?.slot, "env.ANTHROPIC_BASE_URL")
        XCTAssertEqual(plan?.occupied.first?.current, "https://theirs.example")
    }

    /// Occupied is not an outcome. It rides alongside a committable plan,
    /// because IronWire fills the empty slots and reports the full one in the
    /// same pass.
    func testAnOccupiedSlotRidesAlongsideAPlanThatStillHasChanges() {
        let calls = HarnessCalls(
            stateCode: { _ in 31 }, planOutcomeCode: { _ in 41 },
            actionAvailable: { _, _, _ in true })
        let plan = HarnessSurface.plan(
            fromJSON: #"""
                {"id":"claude","action":"connect","outcome":"changes","plan_id":"c0ffee","path":"/p",
                 "changes":["set a thing"],
                 "occupied":[{"slot":"env.ANTHROPIC_BASE_URL","current":"https://theirs.example"}]}
                """#)
        XCTAssertEqual(plan.map { HarnessSurface.canCommit($0, calls: calls) }, true)
        XCTAssertEqual(plan?.occupied.count, 1)
        XCTAssertEqual(
            plan.map { HarnessSurface.outcomeSentence($0, copy: copy(), calls: calls) }, .some(nil))
    }

    /// The words for an occupied slot say it was left alone. They are the
    /// payload's, and the surface has none of its own.
    func testTheOccupiedSentenceIsThePayloadsAndSaysItWasLeftAlone() {
        XCTAssertEqual(HarnessSurface.occupiedSentence(copy: copy()), "H-TAKEN")
    }

    // MARK: - The commit, and a plan that is no longer held

    /// `harness_commit` takes a plan id and nothing else.
    func testTheCommitCarriesTheMintedPlanIdAndNothingElse() {
        let params = HarnessSurface.commitParams(planID: "c0ffee")
        XCTAssertEqual(params.keys.sorted(), ["plan_id"])
        XCTAssertEqual(params["plan_id"] as? String, "c0ffee")
    }

    /// The plan params name a tool and an action, and never a file or a
    /// value to write.
    func testThePlanParamsNameOnlyAToolAndAnAction() {
        let params = HarnessSurface.planParams(id: "claude", action: .connect)
        XCTAssertEqual(params.keys.sorted(), ["action", "id"])
        XCTAssertEqual(params["action"] as? String, "connect")
        XCTAssertEqual(
            HarnessSurface.planParams(id: "claude", action: .disconnect)["action"] as? String,
            "disconnect")
    }

    /// Expired, already committed, never minted, the file moved, the write
    /// failed: the daemon takes the plan out of its store before it checks
    /// any of those, so every one of them leaves nothing to commit again.
    /// None is a retry, and the contributor is told the change did not
    /// happen in the payload's own words.
    func testEveryFailedCommitSpendsThePlanAndIsNotRetried() {
        for code in [
            "harness-plan-unknown", "harness-config-changed", "harness-commit-failed",
            "unavailable",
        ] {
            XCTAssertTrue(
                HarnessSurface.planIsSpent(afterCommitFailure: code),
                "\(code) must not leave a plan id a shell could send again")
        }
        XCTAssertEqual(HarnessSurface.commitFailureSentence(copy: copy()), "UNCONFIRMED")
    }

    /// A connect with nothing answering here is a refusal the daemon makes,
    /// and the shell's answer to it is the exposure question -- not a retry.
    func testAConnectWithNoDestinationIsTheExposureQuestion() {
        XCTAssertTrue(HarnessSurface.isNoDestination("harness-no-destination"))
        XCTAssertFalse(HarnessSurface.isNoDestination("harness-unknown"))
    }

    // MARK: - The exposure gate

    /// The listener is open to everything on this machine, which does not
    /// follow from connecting one tool. Every connect made while it is off
    /// asks first, and that set is a superset of the first-run offer's.
    func testEveryConnectWhileNothingAnswersHereAsksTheExposureQuestion() {
        XCTAssertTrue(HarnessSurface.connectNeedsExposure(listenerOn: false))
        XCTAssertFalse(HarnessSurface.connectNeedsExposure(listenerOn: true))
    }

    /// The gate never lets a first connect past the first-run offer: wherever
    /// the shared table says to ask, the gate asks too.
    func testTheGateIsAtLeastAsCautiousAsTheSharedOfferTable() {
        let offer: @Sendable (Bool, Bool) -> Bool = { !$0 && !$1 }
        for answered in [false, true] {
            for on in [false, true] where offer(answered, on) {
                XCTAssertTrue(HarnessSurface.connectNeedsExposure(listenerOn: on))
            }
        }
    }

    /// Accepting turns the destination on and records the answer in one
    /// write; declining records the answer alone and connects nothing.
    func testAcceptingTurnsItOnAndDecliningWritesTheMarkerAlone() {
        let accept = HarnessSurface.exposureParams(accepted: true)
        XCTAssertEqual(accept["private_inference"] as? Bool, true)
        XCTAssertEqual(accept["private_inference_offer_seen"] as? Bool, true)
        let decline = HarnessSurface.exposureParams(accepted: false)
        XCTAssertNil(decline["private_inference"])
        XCTAssertEqual(decline["private_inference_offer_seen"] as? Bool, true)
    }
}

/// A tiny recorder, so a stub can also be a spy.
private final class SpyBox: @unchecked Sendable {
    private let lock = NSLock()
    private var seen: [String] = []
    func record(_ value: String) {
        lock.lock()
        defer { lock.unlock() }
        seen.append(value)
    }
    var values: [String] {
        lock.lock()
        defer { lock.unlock() }
        return seen
    }
}

File: windows/tests/TraceCommons.Interop.Tests/HarnessSurfaceTests.cs
using System;
using System.Collections.Generic;
using TraceCommons.Interop;
using Xunit;

namespace TraceCommons.Interop.Tests;

/// <summary>
/// The harness list, the plan and the commit, decoded.
///
/// <para>
/// These mirror the macOS shell's cases on purpose: two shells decoding the
/// same payload by different rules would produce two behaviours nobody could
/// compare. Where the plan document's snippets quote an older field set
/// (<c>wired</c>, <c>facades</c>, <c>noop</c>, <c>digest</c>), the cases below
/// quote what <c>daemon/harness.rs</c> actually puts on the wire.
/// </para>
/// </summary>
public class HarnessSurfaceTests
{
    private const string OneRow =
        """
        {"catalog_present":false,"destination_port":8787,
         "harnesses":[{"id":"claude","name":"Claude Code","installed":true,
           "connected":false,"config_path":"C:\\Users\\x\\.claude\\settings.json",
           "connect_command":"ironwire connect claude","family":"anthropic",
           "state":"not_connected","last_call_at":null,
           "can_connect":true,"can_disconnect":false}],
         "activity":{"readable":true,"window_hours":24,"last_call_at":null,"families":[]}}
        """;

    [Fact]
    public void AMalformedRowIsDroppedRatherThanGuessed()
    {
        Assert.Empty(HarnessSurface.ParseListing("""{"harnesses":[{"name":"x"}]}""").Harnesses);
    }

    [Fact]
    public void AMalformedPayloadIsAnEmptyListRatherThanAThrow()
    {
        Assert.Empty(HarnessSurface.ParseListing("{not json").Harnesses);
        Assert.Empty(HarnessSurface.ParseListing(null).Harnesses);
        Assert.Empty(HarnessSurface.ParseListing("  ").Harnesses);
    }

    [Fact]
    public void TheRowCarriesTheFileItWouldChange()
    {
        HarnessListing listing = HarnessSurface.ParseListing(OneRow);
        Assert.Single(listing.Harnesses);
        Assert.Equal(@"C:\Users\x\.claude\settings.json", listing.Harnesses[0].ConfigPath);
        Assert.Equal("ironwire connect claude", listing.Harnesses[0].ConnectCommand);
        Assert.Equal("Claude Code", listing.Harnesses[0].Name);
    }

    /// <summary>
    /// The catalog channel is inert in this build, so the list is the tools
    /// this app was compiled knowing about. A surface that hid that would
    /// imply the machine has only these.
    /// </summary>
    [Fact]
    public void TheListSaysWhetherItCameFromACatalog()
    {
        Assert.False(HarnessSurface.ParseListing(OneRow).CatalogPresent);
    }

    [Fact]
    public void ANoopPlanIsDistinctFromAChange()
    {
        HarnessPlan? noop = HarnessSurface.ParsePlan(
            """{"id":"claude","action":"connect","outcome":"noop","plan_id":null,"path":"/p","changes":[],"occupied":[]}""");
        Assert.Equal(HarnessPlanOutcome.Noop, noop!.Outcome);
        Assert.False(noop.IsCommittable);

        HarnessPlan? real = HarnessSurface.ParsePlan(
            """{"id":"claude","action":"connect","outcome":"changes","plan_id":"a5f0e4a7-2a1f-4f3e-8b1e-0f6c1d2b3a44","path":"/p","changes":["set a thing"],"occupied":[]}""");
        Assert.Equal(HarnessPlanOutcome.Changes, real!.Outcome);
        Assert.True(real.IsCommittable);
        Assert.Equal(new[] { "set a thing" }, real.Changes);
        Assert.Equal("a5f0e4a7-2a1f-4f3e-8b1e-0f6c1d2b3a44", real.PlanId);
    }

    /// <summary>
    /// An unparseable config is a refusal that needs a human, not "nothing to
    /// change". A shell that collapsed the two would tell a contributor with a
    /// broken settings file that everything was fine.
    /// </summary>
    [Fact]
    public void ARefusedFileIsNotANoop()
    {
        HarnessPlan? refused = HarnessSurface.ParsePlan(
            """{"id":"claude","action":"connect","outcome":"unparseable","plan_id":null,"path":"/p","changes":[],"occupied":[]}""");
        Assert.Equal(HarnessPlanOutcome.Unparseable, refused!.Outcome);
        Assert.NotEqual(HarnessPlanOutcome.Noop, refused.Outcome);
        Assert.False(refused.IsCommittable);
    }

    /// <summary>
    /// A plan id is minted for a committable plan and for nothing else, so a
    /// shell cannot construct a write out of an outcome that refused one.
    /// </summary>
    [Fact]
    public void OnlyAChangeCarriesAPlanId()
    {
        foreach (string outcome in new[] { "noop", "unparseable", "not_installed", "entry_unusable", "no_config_path" })
        {
            HarnessPlan? plan = HarnessSurface.ParsePlan(
                $$"""{"id":"claude","action":"connect","outcome":"{{outcome}}","plan_id":null,"path":null,"changes":[],"occupied":[]}""");
            Assert.False(plan!.IsCommittable);
            Assert.Null(plan.PlanId);
        }
    }

    /// <summary>An outcome this build never heard of is not committable.</summary>
    [Fact]
    public void AnOutcomeFromALaterDaemonIsNotCommittable()
    {
        HarnessPlan? plan = HarnessSurface.ParsePlan(
            """{"id":"claude","action":"connect","outcome":"a_thing_invented_later","plan_id":"a5f0e4a7-2a1f-4f3e-8b1e-0f6c1d2b3a44","path":"/p","changes":[],"occupied":[]}""");
        Assert.Equal(HarnessPlanOutcome.Unknown, plan!.Outcome);
        Assert.False(plan.IsCommittable);
    }

    [Fact]
    public void AnOccupiedSlotSurvivesToTheScreen()
    {
        HarnessPlan? plan = HarnessSurface.ParsePlan(
            """{"id":"claude","action":"connect","outcome":"noop","plan_id":null,"path":"/p","changes":[],"occupied":[{"slot":"env.ANTHROPIC_BASE_URL","current":"https://theirs.example"}]}""");
        Assert.Equal("env.ANTHROPIC_BASE_URL", plan!.Occupied[0].Slot);
        Assert.Equal("https://theirs.example", plan.Occupied[0].Current);
    }

    /// <summary>
    /// Occupied rides alongside the outcome; it is not one. A plan can carry
    /// changes AND a slot that was left alone.
    /// </summary>
    [Fact]
    public void AnOccupiedSlotRidesAlongsideAChange()
    {
        HarnessPlan? plan = HarnessSurface.ParsePlan(
            """{"id":"claude","action":"connect","outcome":"changes","plan_id":"a5f0e4a7-2a1f-4f3e-8b1e-0f6c1d2b3a44","path":"/p","changes":["set a thing"],"occupied":[{"slot":"env.ANTHROPIC_BASE_URL","current":"https://theirs.example"}]}""");
        Assert.True(plan!.IsCommittable);
        Assert.Single(plan.Occupied);
    }

    [Fact]
    public void ANotInstalledToolCannotBeConnected()
    {
        var row = new HarnessRow { Id = "codex", Name = "Codex", Installed = false };
        Assert.False(row.CanConnect);
    }

    /// <summary>
    /// Uninstalling a coding tool does not remove the line we put in its
    /// config file, so a connected tool may always be disconnected.
    /// </summary>
    [Fact]
    public void AConnectedToolMayBeDisconnectedEvenUninstalled()
    {
        Assert.True(HarnessSurface.ActionAvailable(HarnessSurface.Disconnect, installed: false, connected: true));
        Assert.False(HarnessSurface.ActionAvailable(HarnessSurface.Connect, installed: false, connected: true));
        Assert.False(HarnessSurface.ActionAvailable("rewire", installed: true, connected: true));
    }

    /// <summary>The daemon's own answer and the ABI's agree, per row.</summary>
    [Fact]
    public void TheRowsAnswerMatchesTheSharedBranchTable()
    {
        HarnessRow row = HarnessSurface.ParseListing(OneRow).Harnesses[0];
        Assert.Equal(
            HarnessSurface.ActionAvailable(HarnessSurface.Connect, row.Installed, row.Connected),
            row.CanConnect);
        Assert.Equal(
            HarnessSurface.ActionAvailable(HarnessSurface.Disconnect, row.Installed, row.Connected),
            row.CanDisconnect);
    }

    /// Only the clear tone may be painted as working, on a harness row as anywhere.
    [Fact]
    public void OnlyAnAnsweringHarnessReadsAsWorking()
    {
        Assert.True(HarnessSurface.ReadsAsWorking(HarnessSurface.State("answering")));
        foreach (string label in new[]
        {
            "not_connected", "connected_no_calls", "activity_shared", "unknown",
            "", "a_state_from_a_later_daemon",
        })
        {
            Assert.False(HarnessSurface.ReadsAsWorking(HarnessSurface.State(label)));
        }
    }

    /// <summary>
    /// A call arrived in this tool's protocol family and more than one
    /// connected tool speaks it. The sentence that says a call was answered
    /// is still true of this computer; what may not happen is painting the
    /// row as this tool working.
    /// </summary>
    [Fact]
    public void ASharedFamilyIsItsOwnStateAndNotAFlavourOfAnswering()
    {
        Assert.Equal(HarnessState.ActivityShared, HarnessSurface.State("activity_shared"));
        Assert.NotEqual(HarnessState.Answering, HarnessSurface.State("activity_shared"));
    }

    [Fact]
    public void AStateFromALaterDaemonIsUnknown()
    {
        Assert.Equal(HarnessState.Unknown, HarnessSurface.State("a_state_from_a_later_daemon"));
        Assert.Equal(HarnessState.Unknown, HarnessSurface.State(null));
    }

    /// <summary>
    /// The shell cannot construct a write. A commit carries the plan id the
    /// daemon minted and nothing else -- no id, no action, no path.
    /// </summary>
    [Fact]
    public void ACommitCarriesOnlyThePlanTheDaemonMinted()
    {
        string body = HarnessSurface.SerializeCommit("a5f0e4a7-2a1f-4f3e-8b1e-0f6c1d2b3a44");
        Assert.Equal("""{"plan_id":"a5f0e4a7-2a1f-4f3e-8b1e-0f6c1d2b3a44"}""", body);
    }

    [Fact]
    public void APlanRequestNamesTheToolAndTheDirection()
    {
        Assert.Equal(
            """{"id":"claude","action":"connect"}""",
            HarnessSurface.SerializePlan("claude", HarnessSurface.Connect));
        Assert.Equal(
            """{"id":"codex","action":"disconnect"}""",
            HarnessSurface.SerializePlan("codex", HarnessSurface.Disconnect));
    }

    /// <summary>
    /// Every commit refusal is the same refusal, and none of them leaves
    /// anything to commit again.
    /// </summary>
    /// <remarks>
    /// The daemon takes the plan out of its store BEFORE it re-checks the
    /// file digest and before it writes, so an expired plan, a spent one, a
    /// file that moved underneath it and a write that failed all end with no
    /// plan and nothing written. A shell that told these apart would be
    /// offering a retry against nothing, or re-planning a write on the
    /// contributor's behalf that they were never shown.
    /// </remarks>
    [Fact]
    public void NoCommitRefusalIsRetryable()
    {
        foreach (string message in new[]
        {
            "harness-plan-unknown", "harness-config-changed", "harness-commit-failed",
            "something-a-later-daemon-says",
        })
        {
            Assert.False(HarnessSurface.CommitIsRetryable(
                new DaemonError { Code = "unavailable", Message = message }));
        }

        Assert.False(HarnessSurface.CommitIsRetryable(null));
    }

    /// <summary>
    /// A connect cannot be planned when nothing answers here. The daemon says
    /// so by name rather than by an outcome, because it is a fact about this
    /// computer and not about the tool's file.
    /// </summary>
    [Fact]
    public void AConnectWithNoDestinationIsRecognised()
    {
        Assert.True(HarnessSurface.NoDestination(new DaemonError { Code = "bad-params", Message = "harness-no-destination" }));
        Assert.False(HarnessSurface.NoDestination(new DaemonError { Code = "bad-params", Message = "harness-unknown" }));
    }

    [Fact]
    public void ACommitAnswerCarriesTheFileAndTheCopyItPreserved()
    {
        HarnessCommit? commit = HarnessSurface.ParseCommit(
            """{"id":"claude","action":"connect","committed":true,"path":"/p","backup_path":"/p.ironwire-backup"}""");
        Assert.True(commit!.Committed);
        Assert.Equal("/p", commit.Path);
        Assert.Equal("/p.ironwire-backup", commit.BackupPath);
        Assert.Null(HarnessSurface.ParseCommit("{nope"));
    }

    /// <summary>
    /// Every sentence a harness row shows comes off the payload. Nothing here
    /// composes one, so an unknown state shows no sentence rather than a
    /// guessed one.
    /// </summary>
    [Fact]
    public void EveryStateSentenceComesFromThePayload()
    {
        PrivateInferenceCopy? copy = PrivateInferenceSurface.Copy();
        Assert.NotNull(copy);

        var expected = new Dictionary<HarnessState, string>
        {
            [HarnessState.NotConnected] = copy!.HarnessNotConnected,
            [HarnessState.ConnectedNoCalls] = copy.HarnessConnectedNothingSeen,
            [HarnessState.Answering] = copy.HarnessAnswering,
        };

        foreach (KeyValuePair<HarnessState, string> pair in expected)
        {
            Assert.Equal(pair.Value, HarnessSurface.StateSentence(pair.Key, copy));
        }
    }

    /// <summary>
    /// Two states have no sentence on the payload, and neither may borrow one.
    /// </summary>
    /// <remarks>
    /// <c>HarnessAnswering</c> would claim an attribution the ledger cannot
    /// make -- it records a protocol family, and a family two connected tools
    /// both speak names neither of them. <c>HarnessConnectedNothingSeen</c>
    /// would be flatly false: something did arrive. So the row draws no state
    /// line, at a tone that is not the working one. This test is what fails if
    /// someone later fills the gap by pointing at the nearest sentence instead
    /// of adding the missing one.
    /// </remarks>
    [Fact]
    public void TheTwoStatesWithNoSentenceBorrowNeither()
    {
        PrivateInferenceCopy? copy = PrivateInferenceSurface.Copy();
        Assert.NotNull(copy);

        foreach (HarnessState state in new[] { HarnessState.ActivityShared, HarnessState.Unknown })
        {
            string sentence = HarnessSurface.StateSentence(state, copy!);
            Assert.Equal(string.Empty, sentence);
            Assert.NotEqual(copy!.HarnessAnswering, sentence);
            Assert.NotEqual(copy.HarnessConnectedNothingSeen, sentence);
            Assert.False(HarnessSurface.ReadsAsWorking(state));
        }
    }

    /// <summary>
    /// A timestamp with fractional seconds does not empty the list.
    /// </summary>
    /// <remarks>
    /// The daemon writes <c>to_rfc3339()</c>, which carries fractional
    /// seconds, and <c>last_call_at</c> sits INSIDE the list envelope -- so a
    /// decoder strict enough to reject it loses every tool, not one field.
    /// This shell carries the value as the string the daemon wrote and never
    /// parses it, because it renders no sentence about when: the one that
    /// exists lives in the Rust and is not exported across the ABI. That makes
    /// this a guard on the decision, not on a parser, and it is the thing that
    /// fails if someone later reaches for a strict date type here.
    /// </remarks>
    [Theory]
    [InlineData("2026-09-07T11:22:33.123456789+00:00")]
    [InlineData("2026-09-07T11:22:33.123456789Z")]
    [InlineData("2026-09-07T11:22:33+00:00")]
    [InlineData("2026-09-07T11:22:33.5-07:00")]
    public void AFractionalSecondTimestampKeepsTheWholeList(string stamp)
    {
        const string template =
            """
            {"catalog_present":false,"destination_port":8787,
             "harnesses":[{"id":"claude","name":"Claude Code","installed":true,
               "connected":true,"config_path":"/p","connect_command":"c",
               "family":"anthropic","state":"answering","last_call_at":"STAMP",
               "can_connect":false,"can_disconnect":true}],
             "activity":{"readable":true,"window_hours":24,
               "last_call_at":"STAMP","families":[]}}
            """;

        IReadOnlyList<HarnessRow> rows =
            HarnessSurface.ParseRows(template.Replace("STAMP", stamp, StringComparison.Ordinal));

        Assert.Single(rows);
        Assert.Equal(stamp, rows[0].LastCallAt);
        Assert.Equal(HarnessState.Answering, rows[0].State);
    }

    /// <summary>
    /// The exposure question is asked off the LISTENER, not off the marker,
    /// and is therefore wider than the first-run offer.
    /// </summary>
    /// <remarks>
    /// A contributor who accepted once and has since used the kill switch is
    /// making the exposure decision afresh: a connect would reopen the
    /// listener they turned off. The first-run offer's own branch stays
    /// exactly as it is -- this is a second, wider question, in the shape the
    /// other two shells use.
    /// </remarks>
    [Fact]
    public void AConnectThatWouldReopenTheListenerAsksAgain()
    {
        Assert.True(HarnessSurface.ConnectNeedsExposure(listenerOn: false));
        Assert.False(HarnessSurface.ConnectNeedsExposure(listenerOn: true));
    }

    /// <summary>
    /// The gate is wider than <c>ShouldOffer</c>, which is why it exists
    /// separately rather than calling it.
    /// </summary>
    [Fact]
    public void TheConnectGateIsWiderThanTheFirstRunOffer()
    {
        // Answered, and then switched off. The first-run offer is done with
        // this contributor; a connect is not.
        Assert.False(PrivateInferenceSurface.ShouldOffer(known: true, answered: true, on: false));
        Assert.True(HarnessSurface.ConnectNeedsExposure(listenerOn: false));
    }
}
